Libc/gen/scandir.3
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)scandir.3	8.1 (Berkeley) 6/4/93
.\" $FreeBSD: src/lib/libc/gen/scandir.3,v 1.9 2007/01/09 00:27:55 imp Exp $
.\"
.Dd May 20, 2008
.Dt SCANDIR 3
.Os
.Sh NAME
.Nm scandir ,
#ifdef UNIFDEF_BLOCKS
.Nm scandir_b ,
#endif
.Nm alphasort
.Nd scan a directory
.Sh SYNOPSIS
.In sys/types.h
.In dirent.h
.Ft int
.Fn scandir "const char *dirname" "struct dirent ***namelist" "int \\*(lp*select\\*(rp\\*(lpstruct dirent *\\*(rp" "int \\*(lp*compar\\*(rp\\*(lpconst void *, const void *\\*(rp"
.Ft int
.Fn alphasort "const void *d1" "const void *d2"
#ifdef UNIFDEF_BLOCKS
.Ft int
.Fn scandir_b "const char *dirname" "struct dirent ***namelist" "int \\*(lp^select\\*(rp\\*(lpstruct dirent *\\*(rp" "int \\*(lp^compar\\*(rp\\*(lpconst void *, const void *\\*(rp"
#endif
.Sh DESCRIPTION
The
.Fn scandir
function
reads the directory
.Fa dirname
and builds an array of pointers to directory
entries using
.Xr malloc 3 .
It returns the number of entries in the array.
A pointer to the array of directory entries is stored in the location
referenced by
.Fa namelist .
.Pp
The
.Fa select
argument is a pointer to a user supplied subroutine which is called by
.Fn scandir
to select which entries are to be included in the array.
The select routine is passed a
pointer to a directory entry and should return a non-zero
value if the directory entry is to be included in the array.
If
.Fa select
is null, then all the directory entries will be included.
.Pp
The
.Fa compar
argument is a pointer to a user supplied subroutine which is passed to
.Xr qsort 3
to sort the completed array.
If this pointer is null, the array is not sorted.
Note that from within the
.Fa compar
subroutine, the two arguments are of type
.Ft const struct dirent ** ,
so that a double-dereference is needed to access the fields in the
.Ft dirent
structure.
.Pp
The
.Fn alphasort
function
is a routine which can be used for the
.Fa compar
argument to sort the array alphabetically.
.Pp
The memory allocated for the array can be deallocated with
.Xr free 3 ,
by freeing each pointer in the array and then the array itself.
#ifdef UNIFDEF_BLOCKS
.Pp
The
.Fn scandir_b
function works the same way as the
.Fn scandir
function, except that
.Fa select
and
.Fa compar
are blocks instead of subroutines.
#endif
.Sh DIAGNOSTICS
Returns \-1 if the directory cannot be opened for reading or if
.Xr malloc 3
cannot allocate enough memory to hold all the data structures.
.Sh SEE ALSO
.Xr directory 3 ,
.Xr malloc 3 ,
.Xr qsort 3 ,
.Xr dir 5
.Sh HISTORY
The
.Fn scandir
and
.Fn alphasort
functions appeared in
.Bx 4.2 .
#ifdef UNIFDEF_BLOCKS
The
.Fn scandir_b
function appeared in Mac OS X 10.6.
#endif


X11/ChangeLog
2006-09-08  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_44

2006-09-06  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.h,
	xc/lib/GL/apple/dri_dispatch.defs: removed glPointParameteri[v]
	aliases

	* xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/applewmstr.h,
	xc/programs/Xserver/hw/apple/applewm.h,
	xc/programs/Xserver/hw/apple/applewm.c,
	xc/programs/Xserver/hw/apple/applewm-impl.h,
	xc/programs/Xserver/hw/apple/applewm-impl.c,
	xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: use shared
	version of AppleWM headers

	* make-host-def, build.sh, Makefile: removed Xplugin/quartz-wm
	references; attempt to support x86_64

2006-07-07  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_43

	* xc/programs/Xserver/hw/apple/x-hash.c: fix compiler warning

	* xc/programs/xterm/main.c: fix xterm build failure (4607761)

2006-02-12  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/GL/apple/indirect.c
	(glAquaInitVisualConfigs): add stereo visuals (4377346)

2006-01-16  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_42

	* xc/programs/Xserver/hw/apple/bundle-main.c: added /sbin and
	/usr/sbin to the default PATH value (4408133)

	* xc/programs/Xserver/hw/apple/quartz.c (do_exec): use same
	default PATH and MANPATH values as bundle-main.c (4408133)

2005-11-29  John Harper  <jsh@apple.com>

	Fix xscreensaver-demo crash and make its cross-process GL
	surfaces work correctly (4356702):

	* xc/programs/Xserver/hw/apple/dri-surface.c: don't compile the
	code to draw on surfaces (that was always disables but was
	causing a crash - I've fixed the crash, but the code is never
	used so best to just remove it from the binary)

2005-10-25  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_41

	* xc/lib/xtrans/Xtranssock.c: fix Leopard build failure
	(4222517)

2005-10-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix default
	MANPATH (4312329)

2005-08-29  John Harper  <jsh@apple.com>

	* xc/extras/freetype2/src/base/ftmac.c: merged endian changes
	from 3477131

2005-06-22  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: merged
	fix for CAN-2005-0605 from xfree86 cvs (4031381)

2005-05-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c
	(install_ipaddr_source): changed the way we listen for network
	/ hostname changes, hopefully to be more reliable

	* xc/programs/Xserver/hw/apple/X11Application.m (sendEvent):
	change fix for bring-all-to-front not to require private
	headers - hardcode enum value we need :-(  (4075093)

2005-05-23  John Harper  <jsh@apple.com>

	* Makefile, make-host-def, merge-libs.sh,
	xc/config/cf/Imake.cf, xc/config/cf/darwin.cf,
	xc/config/cf/darwinLib.tmpl, xc/include/Xmd.h,
	xc/lib/Xt/TMaction.c, xc/lib/Xt/Xtos.h,
	xc/lib/freetype2/Imakefile: checked in changes to
	support building fat libraries when ppc64 is in RC_ARCHS
	(3721701)

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	bumped short version number to 1.2 so we don't fail to verify

	* xc/programs/Xserver/hw/apple/X11Application.m: handle scroll
	wheel events with delta < 1 correctly (4109092)

2005-05-20  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/rootless.h,
	xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Application.m: recognize the
	bfCPSOrderAllWindowsForward bit in activation events and
	respond to it appropriately (4075093)

2005-03-09  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_40

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-03  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_39

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_38

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_37

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-24  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36_1

	* Makefile: strip-libraries target doesn't strip libXplugin

2004-11-23  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)


Libc/gen/FreeBSD/scandir.c
/*
 * Copyright (c) 1983, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)scandir.c	8.3 (Berkeley) 1/2/94";
#endif /* LIBC_SCCS and not lint */
#include <sys/cdefs.h>
__FBSDID("$FreeBSD: src/lib/libc/gen/scandir.c,v 1.9 2008/03/16 19:08:53 das Exp $");

/*
 * Scan the directory dirname calling select to make a list of selected
 * directory entries then sort using qsort and compare routine dcomp.
 * Returns the number of entries and a pointer to a list of pointers to
 * struct dirent (through namelist). Returns -1 if there were any errors.
 */

#include "namespace.h"
#include <dirent.h>
#include <stdlib.h>
#include <string.h>
#include "un-namespace.h"

/*
 * The DIRSIZ macro is the minimum record length which will hold the directory
 * entry.  This requires the amount of space in struct dirent without the
 * d_name field, plus enough space for the name and a terminating nul byte
 * (dp->d_namlen + 1), rounded up to a 4 byte boundary.
 */
#undef DIRSIZ
#define DIRSIZ(dp)							\
	((sizeof(struct dirent) - sizeof(dp)->d_name) +			\
	    (((dp)->d_namlen + 1 + 3) &~ 3))

int
scandir(dirname, namelist, select, dcomp)
	const char *dirname;
	struct dirent ***namelist;
	int (*select)(struct dirent *);
	int (*dcomp)(const void *, const void *);
{
	struct dirent *d, *p, **names = NULL;
	size_t nitems = 0;
	long arraysz;
	DIR *dirp;

	if ((dirp = opendir(dirname)) == NULL)
		return(-1);

	arraysz = 32;	/* initial estimate of the array size */
	names = (struct dirent **)malloc(arraysz * sizeof(struct dirent *));
	if (names == NULL)
		goto fail;

	while ((d = readdir(dirp)) != NULL) {
		if (select != NULL && !(*select)(d))
			continue;	/* just selected names */
		/*
		 * Make a minimum size copy of the data
		 */
		p = (struct dirent *)malloc(DIRSIZ(d));
		if (p == NULL)
			goto fail;
		p->d_fileno = d->d_fileno;
		p->d_type = d->d_type;
		p->d_reclen = d->d_reclen;
		p->d_namlen = d->d_namlen;
		bcopy(d->d_name, p->d_name, p->d_namlen + 1);
		/*
		 * Check to make sure the array has space left and
		 * realloc the maximum size.
		 */
		if (nitems >= arraysz) {
			struct dirent **names2;

			names2 = (struct dirent **)realloc((char *)names,
				(arraysz * 2) * sizeof(struct dirent *));
			if (names2 == NULL) {
				free(p);
				goto fail;
			}
			names = names2;
			arraysz *= 2;
		}
		names[nitems++] = p;
	}
	closedir(dirp);
	if (nitems && dcomp != NULL)
		qsort(names, nitems, sizeof(struct dirent *), dcomp);
	*namelist = names;
	return(nitems);

fail:
	while (nitems > 0)
		free(names[--nitems]);
	free(names);
	closedir(dirp);
	return -1;
}

/*
 * Alphabetic order comparison routine for those who want it.
 */
int
alphasort(d1, d2)
	const void *d1;
	const void *d2;
{
	return(strcmp((*(struct dirent **)d1)->d_name,
	    (*(struct dirent **)d2)->d_name));
}
